### PR TITLE
mrc-2476: Upgrade ajv to version 8.5.0 and include draft-04 extension

### PR DESCRIPTION
**Pull request by r-ash**
See https://ajv.js.org/json-schema.html#draft-04

* Upgrade ajv to 8.5.0
* Include draft-04 schema extension
* Include [ajv-formats plugin](https://github.com/ajv-validator/ajv-formats)
* Add arg `strict` to `json_validator` and `json_validate` FALSE by default which allows permissive handling of schema validation see https://ajv.js.org/strict-mode.html
* If running with schema > draft-04 then fallback to using ajv and print message to user
* Add `dataPath` property to `verbose` errors, for backwards compatibility as this has been renamed to `instancePath` in ajv version 8+

Checked this is working with biocompute, and with hintr. Checked this is working with dependencies using `revdepcheck::revdep_check()` too - failing on TKCat

Issues encountered (which I think I have fixed)
* `TKCat` uses new `2019-09` schema -  `"$schema": "https://json-schema.org/draft/2019-09/schema",`  https://github.com/patzaw/TKCat/blob/master/inst/Collections/Collection-Schema.json#L2 . Ajv supports this see https://ajv.js.org/v6-to-v8-migration.html - tried to work around that by updating the version parsing and making 2019-09 and 2020-12 available
* `dccvalidator` failing as it uses `dataPath` returned in table of errors which has been renamed in `instancePath` in updated ajv, see https://github.com/Sage-Bionetworks/dccvalidator/blob/5dc13f7f77d6444226e3031c1210190937a78682/R/check-schema-json.R#L58 - tried to work around this by duping `dataPath` to `instancePath`
* `tableschema.r` fails if run with ajv because ajv returns 4 errors where imjv returns 1 (we can work around this if we continue to use imjv for draft-04 by default`
* `biocompute` uses `references` which we don't want to fail on e.g. https://github.com/sbg/biocompute/blob/master/inst/schemas/1.3.0-alpha/biocomputeobject.json#L86

Remaining issue
* `TKCat` uses `"$schema": {"enum": ["TKCat_collections_1.0"]},` which throws 
```
Unknown meta schema version 'TKCat_collections_1.0'
E> --- failed re-building ‘TKCat-Collections.Rmd
```
With cran version they are validating against `imjv`. But with updated version currently on master this won't work as it tries to parse the version of the schema so it errors early if not using `draft-04`. This didn't error previously so I assume validated permissively. Now updated to support `2019-09` this is failing validation on ajv. I don't really know what this is - some kind of metaschema? Not sure if `"$schema": {"enum": ["TKCat_collections_1.0"]}` is ever valid?

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+## jsonvalidate 1.3.0
+
+* Upgrade to ajv version 8.5.0
+* Add arg `strict` to `json_validate` and `json_validator` to allow evaluating schema in strict mode for ajv only. This off `FALSE` by default to use permissive behaviour detailed in JSON schema
```
Comment (richfitz) [inline NEWS.md:4]:
```suggestion
* Add arg `strict` to `json_validate` and `json_validator` to allow evaluating schema in strict mode for ajv only. This off (`FALSE`) by default to use permissive behaviour detailed in JSON schema
```

File: R/validate.R [hunk 2]
```diff
@@ -70,12 +78,19 @@ json_validator <- function(schema, engine = "imjv", reference = NULL) {
...
+##'   https://ajv.js.org/strict-mode.html for details. Only available in
+##'   \code{engine = 'ajv'}.
+##'
 ##' @export
 ##' @example man-roxygen/example-json_validate.R
 json_validate <- function(json, schema, verbose = FALSE, greedy = FALSE,
-                          error = FALSE, engine = "imjv", reference = NULL,
-                          query = NULL) {
-  tmp <- json_validator(schema, engine, reference = reference)
+                          error = FALSE, engine = "ajv", reference = NULL,
```
Comment (richfitz) [inline R/validate.R:91]: If we make this change we might break other packages, can you check please?

https://cran.r-project.org/web/packages/jsonvalidate/index.html - see `devtools::revdep` to automate

File: R/validate.R [hunk 3]
```diff
@@ -70,12 +78,19 @@ json_validator <- function(schema, engine = "imjv", reference = NULL) {
...
+                          query = NULL, strict = FALSE) {
+  tmp <- json_validator(schema, engine, reference = reference, strict = strict)
```
Comment (richfitz) [inline R/validate.R:93]: While we're here, could you rename `tmp` to something better (validator, perhaps?)

File: R/validate.R [hunk 4]
```diff
@@ -70,12 +78,19 @@ json_validator <- function(schema, engine = "imjv", reference = NULL) {
 ##'   but for now this must be the name of an element within
 ##'   \code{json}.  See the examples for more details.
 ##'
+##' @param strict Set whether the schema should be parsed strictly or not.
+##'   If in strict mode schemas will error to "prevent any unexpected
+##'   behaviours or silently ignored mistakes in user schema". For example
+##'   it will error if encounters unknown formats or unknown keywords. See
+##'   https://ajv.js.org/strict-mode.html for details. Only available in
```
Comment (richfitz) [inline R/validate.R:85]:
```suggestion
##'   https://ajv.js.org/strict-mode.html for details. Only has an effect when
```

File: js/in.js [hunk 5]
```diff
@@ -1,40 +1,44 @@
...
             .removeKeyword('then')
             .removeKeyword('else');
     } else {
-        var opts = {allErrors: true, verbose: true};
-        var ret = new Ajv({allErrors: true, verbose: true});
+        var ret = new Ajv(opts);
         if (meta_schema_version === "draft-06") {
             ret.addMetaSchema(AjvSchema6);
         }
+        addFormats(ret);
```
Comment (richfitz) [inline js/in.js:33]: Aha, is that how you do this?

File: R/validate.R [hunk 6]
```diff
@@ -33,12 +40,13 @@
...
 ##' @example man-roxygen/example-json_validator.R
-json_validator <- function(schema, engine = "imjv", reference = NULL) {
+json_validator <- function(schema, engine = "ajv", reference = NULL,
+                           strict = FALSE) {
   v8 <- env$ct
   schema <- read_schema(schema, v8)
   switch(engine,
          imjv = json_validator_imjv(schema, v8, reference),
-         ajv = json_validator_ajv(schema, v8, reference),
+         ajv = json_validator_ajv(schema, v8, reference, strict),
```
Comment (richfitz) [inline R/validate.R:49]: What happens if we try and use imjv with a meta schema version of draft-6/7? Still throws an error? We should probably document that. Also the docstring for `engine` needs updating to reflect the new default

**Review by richfitz — CHANGES_REQUESTED** [review]
Nice, this is fantastic. A few minor changes below, most important is to check that we didn't break any of the other packages (if we did we'll need to leave imjv as the default, perhaps unless draft >= 4 is detected)

File: R/validate.R (hunk 6 above)
Comment (r-ash) [reply R/validate.R:49]: It still throws an error yeah - this referenced issue from the code https://github.com/mafintosh/is-my-json-valid/issues/160 still a problem

File: .github/workflows/R-CMD-check.yaml [hunk 7]
```diff
@@ -9,6 +9,7 @@ on:
     branches:
       - main
       - master
+      - mrc-2476
```
Comment (richfitz) [inline .github/workflows/R-CMD-check.yaml:12]:
```suggestion
```

File: .github/workflows/test-coverage.yaml [hunk 8]
```diff
@@ -7,6 +7,7 @@ on:
     branches:
       - main
       - master
+      - mrc-2476
```
Comment (richfitz) [inline .github/workflows/test-coverage.yaml:10]:
```suggestion
```

**Review by richfitz — APPROVED** [review]

**Pull request merged**